src/main/java/com/azure/cosmos/kafka/connect/sink/CosmosDBSinkTask.java

```diff
@@ -93,36 +110,58 @@ public void put(Collection<SinkRecord> records) {
                 }
 
                 maybeInsertId(recordValue, record);
+                toBeWrittenRecordList.add(record);
+            }
+
+            try {
+                SinkWriteResponse response = cosmosdbWriter.write(toBeWrittenRecordList);
 
-                try {
-                    addItemToContainer(container, recordValue);
-                } catch (CosmosException | ConnectException bre) {
-                    sendToDlqIfConfigured(record, bre);
-                } finally {
-                    MDC.clear();
+                if (response.getFailedRecordResponses().size() == 0) {
+                    logger.debug("Sink write completed, {} records succeeded.", response.getSucceededRecords().size());
+
+                } else {
+                    this.sendToDlqIfConfigured(response.getFailedRecordResponses());
                 }
+            }  finally {
+                MDC.clear();
             }
         }
     }
 
+    private boolean isTolerantOnError() {
+        return config.getString(TOLERANCE_ON_ERROR_CONFIG).equalsIgnoreCase("all");
+    }
+
     /**
      * Sends data to a dead letter queue
      *
-     * @param record the kafka record that contains error
+     * @param failedResponses the kafka record that failed to write
      */
-    private void sendToDlqIfConfigured(SinkRecord record, RuntimeException exception) {
-
+    private void sendToDlqIfConfigured(List<SinkOperationFailedResponse> failedResponses) {
         if (context != null && context.errantRecordReporter() != null) {
-            context.errantRecordReporter().report(record, exception);
-            if (!config.getString(TOLERANCE_ON_ERROR_CONFIG).equalsIgnoreCase("all")) {
-                throw exception;
+            for (SinkOperationFailedResponse sinkRecordResponse : failedResponses) {
+                context.errantRecordReporter().report(sinkRecordResponse.getSinkRecord(), sinkRecordResponse.getException());
             }
+        }
+
+        StringBuilder errorMessage = new StringBuilder();
+        for (SinkOperationFailedResponse failedResponse : failedResponses) {
+            errorMessage.append(
+                    String.format(
+                            "Unable to write record to CosmosDB: %s, value schema %s, exception %s",
+                            failedResponse.getSinkRecord().key(),
+                            failedResponse.getSinkRecord().valueSchema(),
+                            failedResponse.getException().getMessage()));
+            errorMessage.append("\n");

```

Comment:
append operation can be chained directly to previous call.

Reply:
updated